mergify_engine/worker.py

```diff
@@ -0,0 +1,366 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import asyncio
+import collections
+import concurrent.futures
+import contextlib
+import datetime
+import functools
+import itertools
+import signal
+import time
+
+import daiquiri
+import msgpack
+import uvloop
+
+from mergify_engine import config
+from mergify_engine import engine
+from mergify_engine import exceptions
+from mergify_engine import utils
+from mergify_engine.clients import github
+
+
+LOG = daiquiri.getLogger(__name__)
+
+
+MAX_RETRIES = 3
+
+WORKER_PROCESSING_DELAY = 30
+
+
+class PullRetry(Exception):
+    def __init__(self, attempts):
+        self.attempts = attempts
+        super().__init__(self)
+
+
+class MaxPullRetry(PullRetry):
+    pass
+
+
+class StreamRetry(Exception):
+    def __init__(self, attempts, retry_at):
+        self.attempts = attempts
+        self.retry_at = retry_at
+        super().__init__(self)
+
+
+class MaxStreamRetry(StreamRetry):
+    pass
+
+
+def push(installation_id, owner, repo, pull_number, event_type, data):
+    redis = utils.get_redis_for_stream()
+    stream_name = f"stream~{installation_id}".encode()
+    scheduled_at = utils.utcnow() + datetime.timedelta(seconds=WORKER_PROCESSING_DELAY)
+    score = scheduled_at.timestamp()
+    transaction = redis.pipeline()
+    # NOTE(sileht): Add this event to the pull request stream
+    transaction.xadd(
+        stream_name,
+        {
+            "event": msgpack.packb(
+                {
+                    "owner": owner,
+                    "repo": repo,
+                    "pull_number": pull_number,
+                    "source": {"event_type": event_type, "data": data},
+                },
+                use_bin_type=True,
+            ),
+        },
+    )
+    # NOTE(sileht): Add pull request stream to process to the list, only if it
+    # does not exists, to not update the score(date)
+    transaction.zadd("streams", {stream_name: score}, nx=True)
+    transaction.execute()
+    LOG.info(
+        "pushed to worker",
+        gh_owner=owner,
+        gh_repo=repo,
+        gh_pull=pull_number,
+        event_type=event_type,
+    )
+
+
+def run_engine(installation_id, owner, repo, pull_number, sources):
+    try:
+        installation = github.get_installation(owner, repo, installation_id)
+    except exceptions.MergifyNotInstalled:
+        return
```

Comment:
do we need to log a warning?

Reply:
It's already logged by the github client. Since this occurs in many place, I was lazy to put the log in all places.